src/java/voldemort/store/readonly/ReadOnlyFileDiffer.java

```diff
@@ -0,0 +1,383 @@
+                oldIndex = oldIter.next();
+                newIndex = newIter.next();
+            } else if (result > 0) {
+                //old hash> new hash.
+                results[ADD]++;
+                newIndex = newIter.next();
+            } else {
+                //old hash < new hash.
+                results[DELETE]++;
+                oldIndex = oldIter.next();
+            }
+        }
+
+        while (oldIndex != null) {
+            results[DELETE]++;
+            oldIndex = oldIter.next();
+        }
+
+        while (newIndex != null) {
+            results[ADD]++;
+            newIndex = newIter.next();
+        }
+        return results;
+    }
+
+    /**
+     * Compare the keys and values one by one between two different versions.
+     */
+    private void compareRecordsWithSameHash(ByteArrayComparator comparator, Pair<byte[], byte[]>[] oldRecords,
+        Pair<byte[], byte[]>[] newRecords, long[] results) {
+        int numOldRecords = oldRecords.length;
+        int numNewRecords = newRecords.length;
+        // These records all have the same hash value, thus maybe there is not fixed order here. \
+        // Just compare them one by one as the number of records should be very small.
+        for (Pair<byte[], byte[]> oldPRecord : oldRecords) {
+            for (Pair<byte[], byte[]> newRecord : newRecords) {
+                if (comparator.compare(oldPRecord.getFirst(), newRecord.getFirst()) == 0) {
+                    numOldRecords--;
+                    numNewRecords--;
+                    if (comparator.compare(oldPRecord.getSecond(), newRecord.getSecond()) == 0) {
+                        results[SAME]++;
+                    } else {
+                        results[UPDATE]++;
+                    }
+                }
+            }
+        }
+        results[ADD] += numNewRecords;
+        results[DELETE] += numOldRecords;
+    }
+
+    static class ByteArrayComparator implements Comparator<byte[]> {
+
+        @Override
+        public int compare(byte[] o1, byte[] o2) {
+            int i = 0;
+            int j = 0;
+            for (; i < o1.length && j < o2.length; i++, j++) {
+                if (o1[i] < o2[i]) {
+                    return -1;
+                } else if (o1[i] > o2[j]) {
+                    return 1;
+                }
+            }
+            if (i < o1.length) {
+                return 1;
+            } else if (j < o2.length) {
+                return -1;
+            } else {
+                return 0;
+            }
+        }
+    }
+
+    /**
+     * This class is used to compare between chunk files. Smaller chunk number means smaller when comparing.
+     */
+    static class ReadOnlyFileNameComparator implements Comparator<String> {
+
+        @Override
+        public int compare(String o1, String o2) {
+            int chunk1 = Integer.valueOf(o1.substring(o1.lastIndexOf('_') + 1));
```

Comment:
The +1 is to tell the ```substring``` function to start its selection at the index which immediately follows the index of the last underscore. In other words, that is the chunk ID in the file name.